src/shared.h

```diff
@@ -302,8 +302,22 @@ to_rtype(Type t)
   return RType(t);
 }
 
+// HACK: Discriminated union workaround
+class Expr;
+
+struct RawExpr {
```

Comment:
Ah, is this just lack of C++11 support for allowing classes as members of unions which have constructors?  In that case, could you rename this class ExprPOD and have a comment explaining that why we need a POD?  Also, minor nits:
 - don't need 'friend' since this is a struct
 - { on new line for class decls.